File: 3º Ano/1º Semestre/SIC/P/src-crypto-signatures/ec_sign.py
#!/usr/bin/python3

import sys
import os
from cryptography.hazmat.primitives.asymmetric import ec, padding, utils
from cryptography.hazmat.primitives import serialization, hashes

def load_ec_priv_key( file ):
    try:
        with open( file, "rb" ) as fd:
            key = serialization.load_pem_private_key( fd.read(), password=None );
    except:
        print( "File %s doesn´t have a valid PEM encoding of a private key" % (file) )
        return None

    if isinstance( key, ec.EllipticCurvePrivateKey ) == False:
        print( "File %s doesn´t have a valid PEM encoding of an EC private key" % (file) )
        return None

    return key

def sign( key, digest_func, fd_in, fd_out ):
    # Since we don't know the size of the input file, we will prehash it in 4K chunks

# TODO: add your code here
# until here
        
    # Now we can sign the digest with ECDSA and the private key provided

    signature = key.sign( digest, ec.ECDSA( utils.Prehashed( digest_func ) ) )
    fd_out.write( signature )

def main( argv ):
    digest_dict = { "SHA-256":hashes.SHA256(), "SHA-384":hashes.SHA384(), "SHA-512":hashes.SHA512() }

    if len(argv) < 3 or argv[2] not in digest_dict:
        print( "Usage: %s priv_file digest_func [file to sign] [signature file]" % (argv[0]) )
        print( "       digest_func can be either SHA-256, SHA-384 or SHA-512" )
        return 0

    pub_key = load_ec_priv_key( argv[1] )

    if pub_key == None:
        return 0

    # If there is a file to sign, use it, otherwise use stdin (reopened in binary mode)

    if len(argv) > 3:
        try:
            fd_in = open( argv[3], "rb" )
        except Exception as e:
            print( "Cannot read from file %s: %s" % (argv[3], e ) )
            return 0
    else:
        fd_in = os.fdopen( sys.stdin.fileno(), "rb" )
    
    # If there is a file to store the signature, use it, otherwise use stdout (reopened in binary mode)

    if len(argv) > 4:
        try:
            fd_out = open( argv[4], "wb" )
        except Exception as e:
            print( "Cannot write to file %s: %s" % (argv[4], e ) )
            return 0
    else:
        fd_out = os.fdopen( sys.stdout.fileno(), "wb" )

    sign( pub_key, digest_dict[argv[2]], fd_in, fd_out )

    return 1

if __name__ == "__main__":
    main( sys.argv )

File: 3º Ano/1º Semestre/SIC/P/src-crypto-signatures/ec_sign.c
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdint.h>
#include <stdint.h>
#include <memory.h>
#include <errno.h>
#include <openssl/pem.h>
#include <openssl/ec.h>
#include <openssl/evp.h>

EVP_PKEY *
load_ec_priv_key( char * file )
{
    FILE * fp = fopen( file, "r" );

    if (fp == 0) {
        printf( "Cannot read file %s: ", file );
	perror( "" );
        printf( "\n");
        return 0;
    }

    EVP_PKEY * key = PEM_read_PrivateKey( fp, &key, 0, 0 );

    if (key == 0) {
        printf( "File %s doesn´t have a valid PEM encoding of a private key\n", file );
        return 0;
    }

    if (EVP_PKEY_get_id( key ) != EVP_PKEY_EC) {
        printf( "File %s doesn´t have a valid PEM encoding of an EC private key\n", file );
        return 0;
    }

    return key;
}

void
sign( EVP_PKEY * key, const EVP_MD * digest_func, int fd_in, int fd_out )
{
    EVP_MD_CTX * md_ctx = EVP_MD_CTX_new();

    // Setup the signature engine, providing only the digest function

    if (EVP_SignInit( md_ctx, digest_func ) == 0) {
        printf( "Internal error, cannot use the provided digest function\n" );
	return;
    }

    // Since we don't know the size of the input file, we will prehash it in 4K chunks
// TODO: add your code here
// until here
        
    // Now we can sign the digest with ECDSA and the private key provided

    if (EVP_SignFinal( md_ctx, signature, &signature_len, key ) == 0) {
        printf( "Internal error, failed to sign\n" );
	return;
    }

    write( fd_out, signature, signature_len );

    // Release dynamic memory

    EVP_MD_CTX_free( md_ctx );
}

struct digest_dict_t {
    char * name;
    const EVP_MD * (*function)();
} digest_dict[4] = {
    {"SHA-256", EVP_sha256},
    {"SHA-384", EVP_sha384},
    {"SHA-512", EVP_sha512},
    {0, 0}
};

void *
get_digest_func( char * name )
{
    for (int i = 0; digest_dict[i].name; i++) {
        if (strcmp( name, digest_dict[i].name ) == 0) {
	    return digest_dict[i].function;
	}
    }

    return 0;
}

int
main( int argc, char * argv[] )
{
    EVP_MD * (*digest_func)();
    int fd_in = 0, fd_out = 1;

    if (argc < 3 || (digest_func = get_digest_func( argv[2] )) == 0) {
        printf( "Usage: %s priv_file digest_func [file to sign] [signature file]\n", argv[0] );
        printf( "       digest_func can be either SHA-256, SHA-384 or SHA-512\n" );
        return 1;
    }

    EVP_PKEY * pub_key = load_ec_priv_key( argv[1] );

    if (pub_key == 0) {
        return 1;
    }

    // If there is a file to sign, use it, otherwise use stdin

    if (argc > 3) {
	fd_in = open( argv[3], 0 );
        if (fd_in == -1) {
            printf( "Cannot read from file %s: ", argv[3] );
	    perror( "" );
	    printf( "\n" );
            return 1;
	}
    }
    
    // If there is a file to store the signature, use it, otherwise use stdout (reopened in binary mode)

    if (argc > 4) {
	fd_out = open( argv[4], 0 );
        if (fd_out == -1) {
            printf( "Cannot write to file %s: ", argv[4] );
	    perror( "" );
	    printf( "\n" );
            return 1;
	}
    }

    sign( pub_key, (*digest_func)(), fd_in, fd_out );

    return 0;
}
